Feature/add tags to recent files

danielronnkvist opened the pull request:
Functionality for adding tags to newly added files.
When files have been chosen and save is clicked, the array is passed via the _$rootScope_ to a new Angular controller.

danielronnkvist committed:
The new method for files is now rendered with angular routes

All routes are specified in the main.coffee file. The base URL is necessary in the application.html.erb, otherwise angular would not recognise the entire url. So I added the @base_href variable to the new methdo so it wont be rendered with that base href all the time.

Moved the templates to other folders, so the new.html.erb file will be a bit easier to read.

danielronnkvist committed:
On save redirect to add TagCtrl

Passing the files to the next controller using the $rootScope.

danielronnkvist committed:
Redirect add tags view to the right controller

danielronnkvist committed:
Input fields for adding tags.

Splits and combines the input fields into arrays for each file.

danielronnkvist committed:
Send the tags to the server

If all tags was successfully added to the server, it redirects.
Since the tags are splitted client side, this isn't done server side.

danielronnkvist committed:
Use the route param instead of copying to a scope param

danielronnkvist committed:
Valdiates that new files have been choosen

danielronnkvist committed:
Added some comments

danielronnkvist committed:
Allways use the base tag

danielronnkvist committed:
The success validation works

nattravn commented:
I did not understand much of the Angularjs part, but everything else looked good. We can probably delete the file views / tags / new.html.erb

klaseskilson commented:
I don't think we should delete `views/tags/new.html.erb`, since that is the only way to add tags to files after they have been added.

nattravn commented:
True! ^^

klaseskilson commented on app/assets/javascripts/controllers/TagCtrl.coffee:
```diff
@@ -0,0 +1,26 @@
+  $scope.files = $rootScope.new_files
+  $scope.tags = ""
+
+  $scope.save = ()->
```
(Minor bullshit comment: [No empty parameter lists!](https://github.com/styleguide/javascript))

klaseskilson commented on app/assets/javascripts/controllers/TagCtrl.coffee:
```diff
@@ -0,0 +1,26 @@
+    Remove empty strings with the _.compact method.
+    And lastly combine two arrays with concat.
+    ###
+    $scope.files.forEach((file)->
```
(Minor bullshit comment number two: [Spaces before function arrows!](https://github.com/styleguide/javascript))

klaseskilson commented on app/assets/javascripts/controllers/TagCtrl.coffee:
```diff
@@ -0,0 +1,26 @@
+          success = data.success
+          )
+      )
+    window.location.replace('/bruse_files') if success
```
Isn't there a risk that this happens before `TagHandler.put(...` is finished? JS won't wait for the asynchronous call to the backend.

danielronnkvist replied:
I could put the loop logic in the TagHandler instead, and then call the redirect in the callback.

klaseskilson commented:
Ruby filenames should be snake_cased to follow the [style guide](https://github.com/bbatsov/ruby-style-guide#naming) on which GitHub's style guide is based.

klaseskilson commented:
I'm not sure I understand what gets prioritized by the system here. If I go to `/service/:identity_id/files/add`, where does it take me? Angular or Rails?

danielronnkvist commented:
`/service/:identity_id/files/add` is redirected to the `views/files/files/new.html.erb`. And that's where Angular takes over the control of the request.

danielronnkvist committed:
Minor syntax bullshit

klaseskilson commented:
Ok. So [this](https://github.com/tkomstadius/bruse/blob/69986ea23b817c7b4c835e1710a7919a2accd843/app/assets/javascripts/main.coffee#L5-L9) get triggered only then? Or why is there duplicate route logic?

danielronnkvist commented:
No. [This](https://github.com/tkomstadius/bruse/blob/69986ea23b817c7b4c835e1710a7919a2accd843/app/assets/javascripts/main.coffee#L10-L14) gets triggered. Since the `<base href="/">` tag is changing the base url of the request to only `/`. I think we have to use duplicate logic to make Rails routing work with Angular routing.

And since the Angular routing is in HTML5 mode the `/service/:identity_id/files/add/tag` renderes the add tags template. But in the `TagsCtrl.coffee`-file I check if their are any files on the `$rootScope`. If not the user is redirected to `/service/:identity_id/files/add`.

Think of it like this; once Rails have rendered the `app/views/files/files/new.html.erb`, Angular takes over the routing control.

klaseskilson commented:
All right! Cool.

danielronnkvist committed:
Recursive function to make sure all tags are added

But it takes a while.. If many files have been added the server is busy serving the request to add files so the add tags function has to wait for its turn.

danielronnkvist committed:
Merge branch 'develop' of github.com:tkomstadius/bruse into feature/add-tags-to-recent-files

Conflicts:
	app/views/files/files/new.html.erb

danielronnkvist commented:
I have fixed the issue with asynchronous javascript. But we have a problem with all the request that are being sent to the server. If I add approx. 200 files, and then add tags to all the files, the system is still working on the add files process, so the request to add tags is delayed.

This problem is probably an effect of the _Fuzzily_ gem, if could postpone the creation of all trigrams I think the problem with the delay will be solved.

klaseskilson commented:
What if we create a bulk method of some sort, so that we can decrease the number of http calls?

So instead of 200 `POST /service/123/files` calls, we use one `POST /service/123/bulk` call.

klaseskilson commented:
Or investigate if the after_save (or whatever hook is beeing used) in Fuzzily can be prevented, and use some sort of delay job?

danielronnkvist commented:
We could do that, but would that solve the problem with how much time Fuzzily takes? There is a method in Fuzzily to create trigrams after an object have been created, so a delayed job is possible.

danielronnkvist committed:
Minor syntax fix

klaseskilson commented:
What if we wrap everything [`ActiveRecord::Transactions`](http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html) then? This would allow the server to only do one db commit (which is basically what takes time) instead of one for each trigram.

This probably has some disadvantages though. For one, there might be a maximum number of SQL commands per commit. I also think that if one query fails, it leads to a total collapse of the entire transaction (:poop:). [Interesting read on the subject!](https://www.coffeepowered.net/2009/01/23/mass-inserting-data-in-rails-without-killing-your-performance/)

Like this:

``` ruby
ActiveRecord::Base.transaction do
  1000.times { Model.create(options) }
end
```

danielronnkvist commented:
A bulk request to the server that will be handled by the `ActiveRecord::Transactions` sounds good. Screw failures, speed over reliability hehe :smirk_cat:

danielronnkvist committed:
New redirect url

danielronnkvist committed:
Merge branch 'develop' of github.com:tkomstadius/bruse into feature/add-tags-to-recent-files

Conflicts:
	app/assets/javascripts/controllers/FileCtrl.coffee
	app/views/files/files/new.html.erb

klaseskilson commented:
Looks good to me!

klaseskilson committed:
Change name of partials in files#new view

danielronnkvist committed:
Redirects when angular shouldn't control the routing

Quick solution and should be looked into more.

danielronnkvist committed:
Same solution as before, just fewer lines

nattravn commented:
The upload function for local files doesn't work now but maybe that isn't a problem if we gonna change the structure in the feature

klaseskilson commented:
I don't think we should approve something that's broken to be merge into develop. Sure, we might change how things work down the line, but that doesn't change what we have today.

danielronnkvist commented:
But is it the file upload or the function to add tags in post that is broken?

nattravn commented:
The upload function/button doesn't work for me anymore. Byebug says that the identity_id from the query is null  ....AND "bruse_files"."identity_id" IS NULL) LIMIT 1. If the identity_id is null it throwing you to the else-state "you must log in to your bruse acount!" and I signed in as a bruse user

nattravn commented:
I had to reset the database and that fixed my problem (I always forget to do that). But the add tag function doesn't work

danielronnkvist committed:
Add tags in post

danielronnkvist commented:
I have fixed the issue with adding tags in post.

I have another solution to the angular routing problem. Instead of having the controller in `otherwise`, we could set all links in the header with the HTML attribute `target="_self"`. So it would look something like this:

``` ruby
<%= link_to 'My files', bruse_files_path, target: "_self"  %>
```

By doing like that the browser is forced to call the server and Angular wont intercept. It will also prevent a small loading time caused by the redirect controller.

danielronnkvist committed:
Added the _self target to links

A combination of this and the redirect controller in Angular will prevent Angular from intercepting wrong request and only the ones specified in `main.coffee`.

klaseskilson commented:
I think it's ready for merge!

nattravn commented:
It works very good! I merge it

johanhenrikssn commented:
Amazing!
